packages/core/core-flows/src/rbac/steps/validate-user-permissions.ts

```diff
@@ -0,0 +1,87 @@
+import {
+  ContainerRegistrationKeys,
+  MedusaError,
+} from "@medusajs/framework/utils"
+import { createStep } from "@medusajs/framework/workflows-sdk"
+
+export type ValidateUserPermissionsStepInput = {
+  user_id: string
+  policy_ids?: string[]
+  actions?: {
+    resource: string
+    operation: string
+  }[]
+}
+
+export const validateUserPermissionsStepId = "validate-user-permissions"
+
+/**
+ * Validates that a user has access to all the policies they are trying to assign.
+ * A user can only create roles and add policies that they themselves have access to.
+ */
+export const validateUserPermissionsStep = createStep(
```

Comment:
thought(non-blocking): realistically, the flow would probably be that certain users, e.g. admins, have the `manage:roles` and `manage:policies` permissions and are therefore the only ones capable of assigning user roles ... don't you think?